config/seata-config-consul/src/main/java/io/seata/config/consul/ConsulConfiguration.java

```diff
@@ -0,0 +1,229 @@
+    }
+
+    @Override
+    public void addConfigListener(String dataId, ConfigChangeListener listener) {
+        configListenersMap.putIfAbsent(dataId, new ArrayList<>());
+        if (null != listener.getExecutor()) {
+            ConfigChangeNotifier configChangeNotifier = new ConfigChangeNotifier(dataId, listener);
+            listener.getExecutor().submit(configChangeNotifier);
+        }
+    }
+
+    @Override
+    public void removeConfigListener(String dataId, ConfigChangeListener listener) {
+        List<ConfigChangeListener> configChangeListeners = getConfigListeners(dataId);
+        if (configChangeListeners == null) {
+            return;
+        }
+        List<ConfigChangeListener> newChangeListenerList = new ArrayList<>();
+        for (ConfigChangeListener changeListener : configChangeListeners) {
+            if (!changeListener.equals(listener)) {
+                newChangeListenerList.add(changeListener);
+            }
+        }
+        configListenersMap.put(dataId, newChangeListenerList);
+        if (null != listener.getExecutor()) {
+            listener.getExecutor().shutdownNow();
+        }
+    }
+
+    @Override
+    public List<ConfigChangeListener> getConfigListeners(String dataId) {
+        return configListenersMap.get(dataId);
+    }
+
+    @Override
+    public String getTypeName() {
+        return REGISTRY_TYPE;
+    }
+
+
+    /**
+     * get consul client
+     *
+     * @return client
+     */
+    private static ConsulClient getConsulClient() {
+        if (null == client) {
+            synchronized (ConsulConfiguration.class) {
+                if (null == client) {
+                    client = new ConsulClient(FILE_CONFIG.getConfig(FILE_CONFIG_KEY_PREFIX + SERVER_ADDR_KEY));
+                }
+            }
+        }
+        return client;
+    }
+
+    /**
+     * complete the future
+     *
+     * @param response
+     * @param configFuture
+     */
+    private void complete(Response response, ConfigFuture configFuture) {
+        Object value = response.getValue();
+        if (null != response && null != value) {
+            if (value instanceof GetValue) {
+                configFuture.setResult(((GetValue) response.getValue()).getDecodedValue());
+            } else {
+                configFuture.setResult(response.getValue());
+            }
+        }
+    }
+
+    /**
+     * the type config change notifier
+     */
+    private static class ConfigChangeNotifier implements Runnable {
+        private final String dataId;
+        private final ConfigChangeListener listener;
+        private long consulIndex;
+
+        public ConfigChangeNotifier(String dataId, ConfigChangeListener listener) {
+            this.dataId = dataId;
+            this.listener = listener;
+            this.consulIndex = getConsulClient().getKVValue(this.dataId).getConsulIndex();
+        }
+
+        @Override
+        public void run() {
+            for (; ; ) {
```

Comment:
why not return or break ?

Reply:
This is a configuration notifier that notifies subscribers when a configuration change occurs. This should be continuous rather than stopping at one notification because the configuration can change many times.